### PR TITLE
Add cache-breaking gravatar notification

### PR DESCRIPTION
**Pull request by nheagy**
This is the main changes to allow gravatar images to be updated in one place and take effect throughout the app.

Mostly: call `updateGravatar(image: UIImage, email: String?)` and it will post a `GravatarImageUpdateNotification` that all other UIImageViews displays the same gravatar will then display.

It's a bit of a hack, but it gets the job done and is _fairly_ easy to understand.

### CODE REVIEW

File: WordPressUI/Extensions/UIImageView+Gravatar.swift [hunk 1]
```diff
@@ -131,23 +171,31 @@ extension UIImageView {
...
 
     /// Returns the required gravatar size. If the current view's size is zero, falls back to the default size.
     ///
     private func gravatarDefaultSize() -> Int {
-        guard bounds.size.equalTo(.zero) == false else {
-            return Defaults.imageSize
-        }
-
-        let targetSize = max(bounds.width, bounds.height) * UIScreen.main.scale
-        return Int(targetSize)
```
Comment (jleandroperez) [inline WordPressUI/Extensions/UIImageView+Gravatar.swift]: Nate, what's the reason behind this change? We were downloading an image to fit the size of the UIImageView. With this change we should see slower downloads, and higher bandwidth usage. Can we revert this one?

File: WordPressUI/Extensions/UIImageView+Networking.swift [hunk 2]
```diff
@@ -128,7 +128,7 @@ public extension UIImageView {
 
     /// Stores the Image's remote URL, if any.
     ///
-    private var downloadURL: URL? {
+    internal var downloadURL: URL? {
```
Comment (jleandroperez) [inline WordPressUI/Extensions/UIImageView+Networking.swift]: By default visibility is `internal`, no need to specify it.

File: WordPressUI/Extensions/UIImageView+Gravatar.swift [hunk 3]
```diff
@@ -51,12 +51,32 @@ extension UIImageView {
...
+                let image = userInfo["image"] as? UIImage,
+                let downloadURL = self?.downloadURL else {
+                return
+            }
+            let testHash = self?.gravatarHash(of: email) ?? ""
+            if downloadURL.absoluteString.contains(testHash) {
+                self?.image = image
+            }
+        }
+    }
```
Comment (jleandroperez) [inline WordPressUI/Extensions/UIImageView+Gravatar.swift]: I'm afraid this might not work. I've recently done something similar, and bottom line is:

- This method returns an opaque object
- As per the docs, you're supposed to keep it around, and [use it to remove the observer](https://developer.apple.com/documentation/foundation/notificationcenter/1411723-addobserver)
- In my actual tests (in another PR!), if you don't retain it, the `observer` block gets removed automatically

Plus: There is no "remove observer" mechanism.

File: WordPressUI/Extensions/UIImageView+Gravatar.swift [hunk 4]
```diff
@@ -110,15 +136,29 @@ extension UIImageView {
     /// P.s.:
     /// Hope buddah, and the code reviewer, can forgive me for this hack.
     ///
-    @objc
-    public func overrideGravatarImageCache(_ image: UIImage, rating: GravatarRatings, email: String) {
-        guard let gravatarURL = gravatarUrl(for: email, size: gravatarDefaultSize(), rating: rating.stringValue()) else {
+    @available(*, deprecated)
```
Comment (jleandroperez) [inline WordPressUI/Extensions/UIImageView+Gravatar.swift]: LOVE the *deprecated* usage

File: WordPressUI/Extensions/UIImageView+Gravatar.swift (hunk 3 above)
Comment (jleandroperez) [reply WordPressUI/Extensions/UIImageView+Gravatar.swift]: Correction: The "dealloc gets me unsubscribed" was actually KVO. But still, we could use a "remove observer" call, otherwise the block will end up lurking indefinitely.

File: WordPressUI/Extensions/UIImageView+Gravatar.swift [hunk 5]
```diff
@@ -110,15 +136,29 @@ extension UIImageView {
...
+        listenForGravatarChanges()
         overrideImageCache(for: gravatarURL, with: image)
     }
 
+    /// Updates the gravatar image for the given email, and notifies all gravatar image views
+    ///
+    /// - Parameters:
+    ///   - image: the new UIImage
+    ///   - email: associated email of the new gravatar
+    @objc public func updateGravatar(image: UIImage, email: String?) {
```
Comment (ScoutHarris) [inline WordPressUI/Extensions/UIImageView+Gravatar.swift]: I'm gonna jump in with a question. I'm curious as to why the `email` is optional. Two lines down the `guard` checks for it, so it _appears_ to be necessary.

File: WordPressUI/Extensions/UIImageView+Gravatar.swift (hunk 1 above)
Comment (nheagy) [reply WordPressUI/Extensions/UIImageView+Gravatar.swift]: For the notification to work, we need to download a size that works in multiple locations. We'd be downloading 80px * display res, which is 240px max. I figured it wasn't a problem. Without this change, the solution becomes more complicated…

File: WordPressUI/Extensions/UIImageView+Gravatar.swift (hunk 5 above)
Comment (nheagy) [reply WordPressUI/Extensions/UIImageView+Gravatar.swift]: I flip-flopped on this: when called in `EpilogueUserInfoCell` the email is optional. It either needs a guard here or there. I thought putting it here makes `@objc` interaction easier, but I could be convinced to move it.

File: WordPressUI/Extensions/UIImageView+Gravatar.swift (hunk 3 above)
Comment (nheagy) [reply WordPressUI/Extensions/UIImageView+Gravatar.swift]: it definitely _is_ working, but I think you're right—there needs to be a remove call somewhere. 👍

File: WordPressUI/Extensions/UIImageView+Networking.swift (hunk 2 above)
Comment (nheagy) [reply WordPressUI/Extensions/UIImageView+Networking.swift]: The extension is `public extension UIImageView` so the default access level for this extension is `public`.

**Comment by nheagy** [comment]
Some changes made, ready for your re-evaluation, @jleandroperez

File: WordPressUI/Extensions/UIImageView+Networking.swift (hunk 2 above)
Comment (jleandroperez) [reply WordPressUI/Extensions/UIImageView+Networking.swift]: TIL that's only valid for extensions. Thanks Nate!!!

File: WordPressUI/Extensions/UIImageView+Gravatar.swift [hunk 6]
```diff
@@ -51,12 +51,49 @@ extension UIImageView {
...
+    /// Configures the UIImageView to listen for changes to the gravatar it is displaying
+    private func listenForGravatarChanges() {
+        guard gravatarObserver == nil else {
+            return
+        }
+
+        gravatarObserver = NotificationCenter.default.addObserver(forName: .GravatarImageUpdateNotification, object: nil, queue: nil) { [weak self] (notification) in
+            guard let userInfo = notification.userInfo,
+                let email = userInfo["email"] as? String,
+                let image = userInfo["image"] as? UIImage,
```
Comment (jleandroperez) [inline WordPressUI/Extensions/UIImageView+Gravatar.swift]: Suggestion: It'd be neat to turn both `email` / `image` into a constant, so that we don't need to have the whole string in two spots (repetition / fragile-ness).

File: WordPressUI/Extensions/UIImageView+Gravatar.swift [hunk 7]
```diff
@@ -110,15 +153,29 @@ extension UIImageView {
...
+    ///
+    /// - Parameters:
+    ///   - image: the new UIImage
+    ///   - email: associated email of the new gravatar
+    @objc public func updateGravatar(image: UIImage, email: String?) {
+        self.image = image
+        guard let email = email, let gravatarURL = gravatarUrl(for: email, size: Defaults.imageSize, rating: .x) else {
+            return
+        }
+        NotificationCenter.default.post(name: .GravatarImageUpdateNotification, object: self, userInfo: ["email": email, "image": image])
```
Comment (jleandroperez) [inline WordPressUI/Extensions/UIImageView+Gravatar.swift]: Same as above, prob. a good idea to turn "email" and "image" into actual constants.

Comment (nheagy) [reply WordPressUI/Extensions/UIImageView+Gravatar.swift]: Done and done!

**Comment by nheagy** [comment]
Thanks @jleandroperez and @ScoutHarris!

**Comment by jleandroperez** [comment]
Anytime Nate!! thanks for making these changes!! 🎉
